clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -476,29 +519,67 @@ private void setProducerIdAndEpoch(ProducerIdAndEpoch producerIdAndEpoch) {
      * would not have any way of knowing this happened. So for the transactional producer, it's best to return the
      * produce error to the user and let them abort the transaction and close the producer explicitly.
      */
-    synchronized void resetIdempotentProducerId() {
+    private synchronized void resetIdempotentProducerId() {
         if (isTransactional())
             throw new IllegalStateException("Cannot reset producer state for a transactional producer. " +
                     "You must either abort the ongoing transaction or reinitialize the transactional producer instead");
         setProducerIdAndEpoch(ProducerIdAndEpoch.NONE);
+        transitionTo(State.UNINITIALIZED);
+    }
+
+    private synchronized void resetSequenceForPartition(TopicPartition topicPartition) {
+        topicPartitionBookkeeper.topicPartitions.remove(topicPartition);
+        this.partitionsWithUnresolvedSequences.remove(topicPartition);
+    }
+
+    private synchronized void resetSequenceNumbers() {
         topicPartitionBookkeeper.reset();
         this.partitionsWithUnresolvedSequences.clear();
-        transitionTo(State.UNINITIALIZED);
     }
 
-    synchronized void resetIdempotentProducerIdIfNeeded() {
-        if (!isTransactional()) {
-            if (shouldResetProducerStateAfterResolvingSequences()) {
-                // Check if the previous run expired batches which requires a reset of the producer state.
+    synchronized void requestEpochBumpForPartition(TopicPartition tp) {
+        epochBumpRequired = true;
+        this.partitionsToRewriteSequences.add(tp);
+    }
+
+    // package-private for unit testing
+    synchronized boolean shouldBumpEpoch() {
+        return epochBumpRequired;
+    }
+
+    synchronized void bumpIdempotentProducerEpoch() {
+        if (isTransactional()) {
+            log.debug("Skipping epoch bump for transactional producer. The epoch will be bumped after the ongoing " +
+                    "transaction is aborted");
+        } else {
+            if (this.producerIdAndEpoch.epoch == Short.MAX_VALUE) {
                 resetIdempotentProducerId();
```

Comment:
Probably worth a debug log message inside `resetIdempotentProducerId` and below when we bump the epoch.